### ISSUE TITLE
valid-jsdoc @return validation doesn't take async into account

### ISSUE THREAD

**Issue by Gawdl3y** [eslint/eslint#7307]
**Tell us about your environment**
- **ESLint Version:** 3.7.1
- **Node Version:** 7.0.0 nightly
- **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

``` json
{
    "extends": "eslint:recommended",
    "env": {
        "node": true,
        "es6": true
    },
    "parserOptions": {
        "ecmaVersion": 2017
    },
    "rules": {
        "no-extra-parens": ["warn", "all", {
            "nestedBinaryExpressions": false
        }],
        "valid-jsdoc": ["error", {
            "requireReturn": false
        }],

        "accessor-pairs": "warn",
        "array-callback-return": "error",
        "complexity": "warn",
        "consistent-return": "error",
        "curly": ["error", "multi-line", "consistent"],
        "dot-location": ["error", "property"],
        "dot-notation": "error",
        "eqeqeq": "error",
        "no-empty-function": "error",
        "no-floating-decimal": "error",
        "no-implied-eval": "error",
        "no-invalid-this": "error",
        "no-lone-blocks": "error",
        "no-multi-spaces": "error",
        "no-new-func": "error",
        "no-new-wrappers": "error",
        "no-new": "error",
        "no-octal-escape": "error",
        "no-return-assign": "error",
        "no-self-compare": "error",
        "no-sequences": "error",
        "no-throw-literal": "error",
        "no-unmodified-loop-condition": "error",
        "no-unused-expressions": "error",
        "no-useless-call": "error",
        "no-useless-concat": "error",
        "no-useless-escape": "error",
        "no-void": "error",
        "no-warning-comments": "warn",
        "wrap-iife": "error",
        "yoda": "error",

        "no-label-var": "error",
        "no-shadow": "error",
        "no-undef-init": "error",

        "callback-return": "error",
        "handle-callback-err": "error",
        "no-mixed-requires": "error",
        "no-new-require": "error",
        "no-path-concat": "error",
        "no-process-env": "error",

        "array-bracket-spacing": "error",
        "block-spacing": "error",
        "brace-style": ["error", "1tbs", { "allowSingleLine": true }],
        "camelcase": "error",
        "comma-dangle": "error",
        "comma-spacing": "error",
        "comma-style": "error",
        "computed-property-spacing": "error",
        "consistent-this": "error",
        "eol-last": "error",
        "func-names": "error",
        "func-style": ["error", "declaration", { "allowArrowFunctions": true }],
        "id-length": ["error", { "exceptions": ["i", "j", "a", "b"] }],
        "indent": ["error", "tab", { "SwitchCase": 1 }],
        "key-spacing": "error",
        "keyword-spacing": ["error", {
            "overrides": {
                "if": { "after": false },
                "for": { "after": false },
                "while": { "after": false },
                "catch": { "after": false },
                "switch": { "after": false }
            }
        }],
        "max-depth": "error",
        "max-len": ["error", 120, 2],
        "max-nested-callbacks": ["error", { "max": 4 }],
        "max-statements-per-line": ["error", { "max": 2 }],
        "new-cap": "error",
        "newline-per-chained-call": ["error", { "ignoreChainWithDepth": 3 }],
        "no-array-constructor": "error",
        "no-bitwise": "warn",
        "no-inline-comments": "error",
        "no-lonely-if": "error",
        "no-mixed-operators": "error",
        "no-multiple-empty-lines": ["error", { "max": 2, "maxEOF": 1, "maxBOF": 0 }],
        "no-new-object": "error",
        "no-spaced-func": "error",
        "no-trailing-spaces": "error",
        "no-unneeded-ternary": "error",
        "no-whitespace-before-property": "error",
        "object-curly-newline": "error",
        "object-curly-spacing": ["error", "always"],
        "operator-assignment": "error",
        "operator-linebreak": ["error", "after"],
        "padded-blocks": ["error", "never"],
        "quote-props": ["error", "as-needed"],
        "quotes": ["error", "single", { "avoidEscape": true, "allowTemplateLiterals": true }],
        "semi-spacing": "error",
        "semi": "error",
        "space-before-blocks": "error",
        "space-before-function-paren": ["error", "never"],
        "space-in-parens": "error",
        "space-infix-ops": "error",
        "space-unary-ops": "error",
        "spaced-comment": "error",
        "unicode-bom": "error",

        "arrow-body-style": "error",
        "arrow-spacing": "error",
        "no-duplicate-imports": "error",
        "no-useless-computed-key": "error",
        "no-useless-constructor": "error",
        "prefer-arrow-callback": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "prefer-template": "error",
        "rest-spread-spacing": "error",
        "template-curly-spacing": "error",
        "yield-star-spacing": "error"
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

``` js
/**
 * Runs the command
 * @param {Message} message - The message the command is being run for
 * @param {string[]} args - The arguments for the command, or the matches from a pattern
 * @param {boolean} fromPattern - Whether or not the command is being run from a pattern match or not
 * @return {Promise<CommandResult|string[]|string>} The result of running the command
 */
async run(message, args, fromPattern) { // eslint-disable-line no-unused-vars
    throw new Error(`${this.constructor.name} doesn't have a run() method.`);
}
```

**What did you expect to happen?**
The `@return` tag should accept returning Promises from async functions, regardless of the presence of a `return` statement. All async functions return Promises, always.

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint outputs the error, `Unexpected @return tag; function has no return statement.`.

**Comment by not-an-aardvark** [comment]
Thanks for the report. I'm not convinced this is a bug -- the error message is technically correct in saying that the function has no return statement, and an async function without a return statement will always return `Promise<void>` anyway (as opposed to `Promise<somethingElse>`, like you have in your example). It's possible having an error here would point out cases where someone meant to have a return statement, but didn't.

However, let's leave the issue open and see what the rest of the team thinks.

**Comment by Gawdl3y** [comment]
You would still need to specify that it returns a Promise for the documentation, though, since there aren't any tags for async functions specifically. Even specifying that it returns `Promise<void>` will result in this error.

**Comment by vitorbal** [comment]
I am also not convinced this is a bug – if anything, this might be a feature request for jsdoc, since it appears there [is still no support for Promises and async functions](https://github.com/jsdoc3/jsdoc/issues/509).

As an escape-hatch, we could check if a function is async and simply not report if a `@return` tag is present in that case? Or check against `Promise<void>`, as @not-an-aardvark mentions.

**Comment by mysticatea** [comment]
`valid-jsdoc` rule handles `@returns {void}`/`@returns {undefined}` as nohting.
Yes, I think the rules should behave the same for the following cases:

- `@returns {Promise<void>}`/`@returns {Promise<undefined>}` for async functions.
- `@returns {Generator<void>}`/`@returns {Generator<undefined>}` for generator functions. ([Generator](http://www.ecma-international.org/ecma-262/7.0/#sec-generator-objects))

**Comment by vitorbal** [comment]
@mysticatea Thanks for the reply! In that case I think we should mark this as an "enhancement".

- When function is async, `@returns {Promise<void>}/@returns {Promise<undefined>}` means the function does not return something.
- When function is generator, `@returns {Generator<void>}/@returns {Generator<undefined>}` means the function does not return something.

I think we simply have to enhance the `isValidReturnType` function to check these cases.
@Gawdl3y is this something you would be willing to submit a PR for?

**Comment by not-an-aardvark** [comment]
Should we also make `@returns {undefined}` always be an error for async functions, since they cannot return undefined?

**Comment by vitorbal** [comment]
@not-an-aardvark that's a good point. I think we should check for `{void}` as well, no?

**Comment by mysticatea** [comment]
Yes, that modification is simple on [isValidReturnType](https://github.com/eslint/eslint/blob/cd72bba8449718c7fa4d0f73ffcfe867539df941/lib/rules/valid-jsdoc.js#L126-L128).
The border between enhancements and bugs is sometimes unclear to me :worried: 

@not-an-aardvark good point, but we should be careful if we introduce new warnings. I think it needs new option for the type check.

**Comment by kaicataldo** [comment]
Closing this as discussion has stalled and it doesn't appear there is enough interest, either from the team or from the community, to continue the discussion at this time. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. Please feel free to open a new issue with a proposal to continue this discussion if this something you would like to continue advocating for!

**Comment by morficus** [comment]
Would anyone be interested in re-opening this?
I think what @vitorbal lays out would be great for async and generator functions.

**Comment by basicdays** [comment]
I'd be a fan of that. I have `return undefined; //hack valid-jsdoc` in a lot of my async functions just so I can still have jsdoc linting but to suppress this error.

**Comment by vitorbal** [comment]
@morficus @basicdays thanks for the interest, guys. Would either of you be interested in working on a PR for this if it gets accepted? If so, I can try to muster some support for this change.

**Comment by basicdays** [comment]
I'd certainly be interested. Will need to budget time to get into it. Hopefully by end of month if nobody beats me to it.

**Comment by kaicataldo** [comment]
Feel free to be open a new issue so that we can re-evaluate. Old, closed issues get lost in the backlog and don't get much visibility.
